xchange-poloniex2/src/main/java/info/bitrich/xchangestream/poloniex2/PoloniexStreamingMarketDataService.java

```diff
@@ -39,22 +46,21 @@ public PoloniexStreamingMarketDataService(PoloniexStreamingService service, Map<
     @Override
     public Observable<OrderBook> getOrderBook(CurrencyPair currencyPair, Object... args) {
         Observable<PoloniexOrderbook> subscribedOrderbook = service.subscribeCurrencyPairChannel(currencyPair)
-                .filter(s -> s.getEventType().equals("i") || s.getEventType().equals("o"))
-                .map(s -> {
-                    PoloniexOrderbook orderbook;
-                    if (s.getEventType().equals("i")) {
-                        OrderbookInsertEvent insertEvent = ((PoloniexWebSocketOrderbookInsertEvent) s).getInsert();
-                        SortedMap<BigDecimal, BigDecimal> asks = insertEvent.toDepthLevels(OrderbookInsertEvent.ASK_SIDE);
-                        SortedMap<BigDecimal, BigDecimal> bids = insertEvent.toDepthLevels(OrderbookInsertEvent.BID_SIDE);
-                        orderbook = new PoloniexOrderbook(asks, bids);
-                        orderbooks.put(currencyPair, orderbook);
-                    } else {
-                        OrderbookModifiedEvent modifiedEvent = ((PoloniexWebSocketOrderbookModifiedEvent) s).getModifiedEvent();
-                        orderbook = orderbooks.get(currencyPair);
-                        orderbook.modify(modifiedEvent);
-                    }
-                    return orderbook;
-                });
+                .scan(
+                        Optional.empty(),
+                        (Optional<PoloniexOrderbook> orderbook, List<PoloniexWebSocketEvent> poloniexWebSocketEvents) ->
+                                poloniexWebSocketEvents.stream()
+                                        .filter(s -> s.getEventType().equals("i") || s.getEventType().equals("o"))
```

Comment:
```
.filter(s - > PoloniexWebSocketOrderbookInsertEvent.class.isInstance(s) || PoloniexWebSocketOrderbookModifiedEvent.class.isInstance(s))
```
would be clearer I think

Reply:
True, but I've used `instanceof`, IntelliJ did not like this usage of `class.isInstance`

Reply:
One day I will try and understand IntelliJ. For now I consider it voodoo and stick to Eclipse like a caveman.